crates/sdk/src/websocket.rs

```diff
@@ -287,18 +290,57 @@ impl WsConnection {
                 .observe(msg_size as f64);
         };
 
+        // There is a small but plausible chance that a client's socket will not
+        // be notified that the remote end has closed the connection, e.g.
+        // because of the remote machine being power cycled, or middleboxes
+        // misbehaving.
+        //
+        // Unless the client uses dynamic subscriptions, it will only ever try
+        // to read from the socket, and thus not notice the connection closure.
+        //
+        // For certain types of clients it is crucial to eventually time out
+        // such connections, and attempt to reconnect. We don't, however, want
+        // to flood the server with `Ping` frames unnecessarily.
+        //
+        // Instead, we:
+        //
+        // * Check every `IDLE_TIMEOUT` whether some data has arrived.
+        //
+        //   - If not, send a `Ping` frame.
+        //
+        // * Check after another `IDLE_TIMEOUT` whether data has arrived.
+        //
+        //   - If not, and we were expecting a `Pong` response, consider the
+        //     connection bad and exit the loop, thereby closing the socket.
+        //
+        // Note that the server also initiates `Ping`s, currently at `2 * IDLE_TIMEOUT`.
+        // If both ends cannot communicate, we assume the server has already
+        // timed out the client, and so don't bother sending a `Close` frame.
+        const IDLE_TIMEOUT: Duration = Duration::from_secs(30);
+        let mut idle_timeout_interval = tokio::time::interval_at(Instant::now() + IDLE_TIMEOUT, IDLE_TIMEOUT);
+
+        let mut idle = true;
+        let mut want_pong = false;
+
         let mut outgoing_messages = Some(outgoing_messages);
         loop {
             tokio::select! {
                 incoming = self.sock.try_next() => match incoming {
-                    Err(tokio_tungstenite::tungstenite::error::Error::ConnectionClosed) | Ok(None) => break,
+                    Err(tokio_tungstenite::tungstenite::error::Error::ConnectionClosed) | Ok(None) => {
+                        log::info!("Connection closed");
+                        break;
+                    },
 
-                    Err(e) => Self::maybe_log_error::<(), _>(
-                        "Error reading message from read WebSocket stream",
-                        Err(e),
-                    ),
+                    Err(e) => {
+                        Self::maybe_log_error::<(), _>(
+                            "Error reading message from read WebSocket stream",
+                            Err(e),
+                        );
+                        break;
```

Comment:
None of the other `Error` variants indicate a particularly recoverable situation, so I added this `exit`.